### PR TITLE
fix(fetch): fix ReadableStream memory leak when using stream body

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Summary

This PR fixes a memory leak that occurs when `fetch()` is called with a `ReadableStream` body. The ReadableStream objects were not being properly released, causing them to accumulate in memory.

## Problem

When using `fetch()` with a ReadableStream body:

```javascript
const stream = new ReadableStream({
  start(controller) {
    controller.enqueue(new TextEncoder().encode("data"));
    controller.close();
  }
});

await fetch(url, { method: "POST", body: stream });
```

The ReadableStream objects leak because `FetchTasklet.clearData()` has a conditional that prevents `detach()` from being called on ReadableStream request bodies after streaming has started.

### Root Cause

The problematic condition in `clearData()`:

```zig
if (this.request_body != .ReadableStream or this.is_waiting_request_stream_start) {
    this.request_body.detach();
}
```

After `startRequestStream()` is called:
- `is_waiting_request_stream_start` becomes `false`
- `request_body` is still `.ReadableStream`
- The condition evaluates to `(false or false) = false`
- `detach()` is skipped → **memory leak**

### Why the Original Code Was Wrong

The original code appears to assume that when `startRequestStream()` is called, ownership of the Strong reference is transferred to `ResumableSink`. However, this is incorrect:

1. `startRequestStream()` creates a **new independent** Strong reference in `ResumableSink` (see `ResumableSink.zig:119`)
2. The FetchTasklet's original reference is **not transferred** - it becomes redundant
3. Strong references in Bun are independent - calling `deinit()` on one does not affect the other

## Solution

Remove the conditional and always call `detach()`:

```zig
// Always detach request_body regardless of type.
// When request_body is a ReadableStream, startRequestStream() creates
// an independent Strong reference in ResumableSink, so FetchTasklet's
// reference becomes redundant and must be released to avoid leaks.
this.request_body.detach();
```

### Safety Analysis

This change is safe because:

1. **Strong references are independent**: Each Strong reference maintains its own ref count. Detaching FetchTasklet's reference doesn't affect ResumableSink's reference
2. **Idempotency**: `detach()` is safe to call on already-detached references
3. **Timing**: `clearData()` is only called from `deinit()` after streaming has completed (ref_count = 0)
4. **No UAF risk**: `deinit()` only runs when ref_count reaches 0, which means all streaming operations have completed

## Test Results

Before fix (with system Bun):
```
Expected: <= 100
Received: 501   (Request objects leaked)
Received: 1002  (ReadableStream objects leaked)
```

After fix:
```
6 pass
0 fail
```

## Test Coverage

Added comprehensive tests in `test/js/web/fetch/fetch-cyclic-reference.test.ts` covering:
- Response stream leaks with cyclic references
- Streaming response body leaks
- Request body stream leaks with cyclic references
- ReadableStream body leaks (no cyclic reference needed to reproduce)
- Concurrent fetch operations with cyclic references

### CODE REVIEW

File: test/js/web/fetch/fetch-cyclic-reference.test.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
+import { heapStats } from "bun:jsc";
+import { afterAll, describe, expect, test } from "bun:test";
+
+describe("FetchTasklet cyclic reference", () => {
+  let server: ReturnType<typeof Bun.serve> | null = null;
+
+  afterAll(() => {
```
Comment (Jarred-Sumner) [inline test/js/web/fetch/fetch-cyclic-reference.test.ts]: This is not necessary

File: test/js/web/fetch/fetch-cyclic-reference.test.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    server?.stop(true);
+  });
+
+  test("response stream should not leak when response has cyclic reference", async () => {
+    server = Bun.serve({
```
Comment (Jarred-Sumner) [inline test/js/web/fetch/fetch-cyclic-reference.test.ts]:
```suggestion
    await using server = Bun.serve({
```

File: test/js/web/fetch/fetch-cyclic-reference.test.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    for (let i = 0; i < 1000; i++) {
+      await leak();
+    }
+
+    await Bun.sleep(10);
+    Bun.gc(true);
+    await Bun.sleep(10);
+    Bun.gc(true);
+
+    const responseCount = heapStats().objectTypeCounts.Response || 0;
```
Comment (Jarred-Sumner) [inline test/js/web/fetch/fetch-cyclic-reference.test.ts]: It causes the `Response` to be leaked? Not the `ReadableStream`?

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Only two of these tests fail on main.

Let's only keep the failing tests:
```ts
test/js/web/fetch/fetch-cyclic-reference.test.ts:
✓ FetchTasklet cyclic reference > response stream should not leak when response has cyclic reference [74.37ms]
✓ FetchTasklet cyclic reference > response stream should not leak when streaming response body with cyclic reference [54.61ms]
✓ FetchTasklet cyclic reference > response should not leak when body stream references response [56.90ms]
177 |     await Bun.sleep(10);
178 |     Bun.gc(true);
179 |
180 |     const requestCount = heapStats().objectTypeCounts.Request || 0;
181 |     const readableStreamCount = heapStats().objectTypeCounts.ReadableStream || 0;
182 |     expect(requestCount).toBeLessThanOrEqual(100);
                               ^
error: expect(received).toBeLessThanOrEqual(expected)

Expected: <= 100
Received: 501

      at <anonymous> (/Users/jarred/Code/bun/test/js/web/fetch/fetch-cyclic-reference.test.ts:182:26)
✗ FetchTasklet cyclic reference > fetch with request body stream should not leak with cyclic reference [63.52ms]
220 |     await Bun.sleep(10);
221 |     Bun.gc(true);
222 |
223 |     const readableStreamCount = heapStats().objectTypeCounts.ReadableStream || 0;
224 |     // This currently fails with ~502 streams leaked
225 |     expect(readableStreamCount).toBeLessThanOrEqual(100);
                                      ^
error: expect(received).toBeLessThanOrEqual(expected)

Expected: <= 100
Received: 1002

      at <anonymous> (/Users/jarred/Code/bun/test/js/web/fetch/fetch-cyclic-reference.test.ts:225:33)
✗ FetchTasklet cyclic reference > fetch with ReadableStream body should not leak streams [59.81ms]
✓ FetchTasklet cyclic reference > multiple concurrent fetches should not leak with cyclic references [40.42ms]
```
